File: Doubly Linked List/doubly_Linked_List.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }
  push(val) {
    let newNode = new Node(val);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      newNode.prev = this.tail;
      this.tail = newNode;
    }
    this.length++;
    return newNode;
  }
  pop() {
    if (!this.tail) return undefined;
    let current = this.tail;
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      this.tail = current.prev;
      this.tail.next = null;
      current.prev = null;
    }
    this.length--;
    return current;
  }
  shift() {
    if (!this.head) return undefined;
    let current = this.head;
    this.head = current.next;
    this.head.prev = null;
    current.next = null;
    current.prev = null;
    this.length--;
    return current;
  }
  unshift(val) {
    let newNode = new Node(val);
    if (!this.head) {
      this.push(val);
    } else {
      let currentNext = this.head;
      this.head = newNode;
      this.head.next = currentNext;
      this.length++;
      return newNode;
    }
  }
  getValByIndex(index) {
    if (!this.head) return undefined;
    if (index < 0 || index >= this.length) return "Index Out of Range";
    if (index === 0) return this.head;
    let current = this.head;
    let count = 0;
    while (current.next) {
      current = current.next;
      count++;
      if (count === index) {
        return current;
      }
    }
  }
  setValByIndex(index, val) {
    let requestedIndex = this.getValByIndex(index);
    if (!requestedIndex) return undefined;
    requestedIndex.val = val;
    return requestedIndex;
  }
}

const d = new DoublyLinkedList();

d.push(1);
d.push(2);
d.push(3);
d.push(4);
d.pop();
d.push(4);
d.shift();
d.unshift(5);
console.log(d.getValByIndex(0));
console.log(d.setValByIndex(0, 1));

File: Linked List/Doubly Linked List/doubly_linked_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def insert(self, val):
        newNode = Node(val)
        if not self.head:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            newNode.prev = self.tail
            self.tail = newNode
        self.length += 1
        return f'------------\nHEAD : {self.head.val} \nTAIL : {self.tail.val}'

    def remove(self):
        if self.head == None:
            return None
        current = self.tail
        removedEL = current.val
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = current.prev
            self.tail.next = None
            current = None
        self.length -= 1
        return removedEL

    def searchIndex(self, val):
        if not self.head:
            return None
        current = self.head
        count = 0
        if current.val == val:
            return f'Found At Index {count}'
        while current.next:
            current = current.next
            count += 1
            if current.val == val:
                return f'Found At Index {count}'
            elif not current.next:
                return "Not Found"


d = DoublyLinkedList()
d.insert(1)
d.insert(2)
d.insert(3)
d.remove()
d.remove()
d.remove()
d.insert(3)
d.insert(2)
d.insert(1)
d.insert(0)

print(d.searchIndex(3))
print(d.searchIndex(2))
print(d.searchIndex(1))
print(d.searchIndex(0))
print(d.searchIndex(10))
